torch/fx/graph.py

```diff
@@ -104,25 +104,35 @@ def create_name(self, candidate: str, obj: Optional[Any]) -> str:
         if obj is not None and obj in self._obj_to_name:
             return self._obj_to_name[obj]
 
-        # delete all characters that are illegal in a Python identifier
-        candidate = re.sub('[^0-9a-zA-Z_]+', '_', candidate)
+        # Normalize `candidate` to be a valid Python identifier.
+        name_buffer = io.StringIO()
+        # 1. Python identifiers can't start with a number
         if candidate[0].isdigit():
-            candidate = f'_{candidate}'
-
-        while candidate in self._used_names or self._is_illegal_name(candidate, obj):
-            match = re.match(r"(.*)_(\d+)$", candidate)
-            if match is None:
-                candidate = candidate + '_1'
+            name_buffer.write('_')
+        # 2. Replace any special characters with `_`
+        for c in candidate:
+            if not c.isalnum():
+                name_buffer.write('_')
             else:
-                base, num = match.group(1, 2)
-                candidate = f'{base}_{int(num) + 1}'
+                name_buffer.write(c)
+
+        base_name = name_buffer.getvalue()
+        if base_name in self._used_names or self._is_illegal_name(base_name, obj):
```

Comment:
there is a bug here, if `foo_1` is an independently used name we're not checking it for collision because `self._used_names` only stores the base name :(